### PR TITLE
Annotate implementations of IErrorTypeSymbol

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for review

File: src/Compilers/CSharp/Portable/Symbols/ErrorTypeSymbol.ErrorTypeParameterSymbol.cs [hunk 1]
```diff
@@ -161,15 +163,15 @@ public override int GetHashCode()
...
+            internal override bool Equals(TypeSymbol? t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverrideOpt = null)
             {
                 if (ReferenceEquals(this, t2))
                 {
                     return true;
                 }
 
                 var other = t2 as ErrorTypeParameterSymbol;
-                return (object)other != null &&
+                return (object?)other != null &&
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ErrorTypeSymbol.ErrorTypeParameterSymbol.cs]: >other [](start = 32, length = 5)

nit: consider using `is object`

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs [hunk 2]
```diff
@@ -31,14 +33,14 @@ internal ExtendedErrorTypeSymbol(CSharpCompilation compilation, string name, int
...
+        internal ExtendedErrorTypeSymbol(NamespaceOrTypeSymbol? containingSymbol, string name, int arity, DiagnosticInfo? errorInfo, bool unreported = false, bool variableUsedBeforeDeclaration = false)
         {
-            Debug.Assert(((object)containingSymbol == null) ||
+            Debug.Assert(((object?)containingSymbol == null) ||
                 (containingSymbol.Kind == SymbolKind.Namespace) ||
                 (containingSymbol.Kind == SymbolKind.NamedType) ||
                 (containingSymbol.Kind == SymbolKind.ErrorType));
 
-            Debug.Assert(name != null);
+            RoslynDebug.Assert(name != null);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: >RoslynDebug [](start = 12, length = 11)

Was this change necessary?

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs [hunk 3]
```diff
@@ -89,15 +91,15 @@ internal ExtendedErrorTypeSymbol AsUnreported()
         private static ImmutableArray<Symbol> UnwrapErrorCandidates(ImmutableArray<Symbol> candidateSymbols)
         {
             var candidate = candidateSymbols.IsEmpty ? null : candidateSymbols[0] as ErrorTypeSymbol;
-            return ((object)candidate != null && !candidate.CandidateSymbols.IsEmpty) ? candidate.CandidateSymbols : candidateSymbols;
+            return ((object?)candidate != null && !candidate.CandidateSymbols.IsEmpty) ? candidate.CandidateSymbols : candidateSymbols;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: >candidate [](start = 29, length = 9)

nit: `is` check instead?

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs [hunk 4]
```diff
@@ -217,7 +219,7 @@ internal override ImmutableArray<NamedTypeSymbol> GetDeclaredInterfaces(ConsList
         /// However, the real definition is in this class so that it can access the private field 
         /// nonErrorGuessType.
         /// </remarks>
-        internal static TypeSymbol ExtractNonErrorType(TypeSymbol oldSymbol)
+        internal static TypeSymbol? ExtractNonErrorType(TypeSymbol oldSymbol)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: >oldSymbol [](start = 67, length = 9)

nullable?

File: src/Compilers/CSharp/Portable/Symbols/MissingMetadataTypeSymbol.cs [hunk 5]
```diff
@@ -26,10 +28,10 @@ internal abstract class MissingMetadataTypeSymbol : ErrorTypeSymbol
         protected readonly int arity;
         protected readonly bool mangleName;
 
-        private MissingMetadataTypeSymbol(string name, int arity, bool mangleName, TupleExtraData tupleData = null)
+        private MissingMetadataTypeSymbol(string name, int arity, bool mangleName, TupleExtraData? tupleData = null)
             : base(tupleData)
         {
-            Debug.Assert(name != null);
+            RoslynDebug.Assert(name != null);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/MissingMetadataTypeSymbol.cs]: >RoslynDebug [](start = 12, length = 11)

nit: seems unnecessary here too

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: Yes. The netstandard version of `Debug` lacks the `[DoesNotReturnIf(false)]` attribute hence it would cause the compiler to believe that `name` was `null` going forward and invalidate the `_name = name` assignment.

---
In reply to: [364482752](https://github.com/dotnet/roslyn/pull/39192#discussion_r364482752) [](ancestors = 364482752)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: Ah, forgot about that. Asserting would introduce a maybe-null state :P

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: I intentionally avoided extraneous code style changes.

File: src/Compilers/CSharp/Portable/Symbols/MissingMetadataTypeSymbol.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/MissingMetadataTypeSymbol.cs]: ➡️ See https://github.com/dotnet/roslyn/pull/39192#discussion_r364485103

File: src/Compilers/CSharp/Portable/Symbols/ErrorTypeSymbol.ErrorTypeParameterSymbol.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/ErrorTypeSymbol.ErrorTypeParameterSymbol.cs]: ➡️ See https://github.com/dotnet/roslyn/pull/39192#discussion_r364483031

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: @sharwell making sure you see this one.

---
In reply to: [364483268](https://github.com/dotnet/roslyn/pull/39192#discussion_r364483268) [](ancestors = 364483268)

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs [hunk 6]
```diff
@@ -281,29 +283,29 @@ internal static TypeKind ExtractNonErrorTypeKind(TypeSymbol oldSymbol)
             return commonTypeKind;
         }
 
-        internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt = null)
+        internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverrideOpt = null)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: >TypeSymbol t2 [](start = 38, length = 13)

Should probably annotate here as well.

Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: Updated

File: src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs (hunk 4 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/ExtendedErrorTypeSymbol.cs]: Updated

**Comment by sharwell** [comment]
@333fred @jcouv for second review
